snuba/datasets/groups.py

```diff
@@ -0,0 +1,124 @@
+
+    EVENTS_ALIAS = "events"
+    GROUPS_ALIAS = "groups"
+
+    QUALIFIED_COLUMN_REGEX = re.compile(r"^([a-zA-Z_][a-zA-Z0-9_]*)\.([a-zA-Z0-9_\.\[\]]+)$")
+
+    def __init__(self) -> None:
+        self.__grouped_message = get_dataset("groupedmessage")
+        self.__events = get_dataset("events")
+
+        join_structure = JoinStructure(
+            left_source=SchemaJoinedSource(
+                self.GROUPS_ALIAS,
+                self.__grouped_message.get_dataset_schemas().get_read_schema(),
+            ),
+            right_source=SchemaJoinedSource(
+                self.EVENTS_ALIAS,
+                self.__events.get_dataset_schemas().get_read_schema(),
+            ),
+            mapping=[
+                JoinCondition(
+                    left=JoinConditionExpression(
+                        table_alias=self.GROUPS_ALIAS,
+                        column="project_id"),
+                    right=JoinConditionExpression(
+                        table_alias=self.EVENTS_ALIAS,
+                        column="project_id"),
+                ),
+                JoinCondition(
+                    left=JoinConditionExpression(
+                        table_alias=self.GROUPS_ALIAS,
+                        column="id"),
+                    right=JoinConditionExpression(
+                        table_alias=self.EVENTS_ALIAS,
+                        column="group_id"),
+                ),
+            ],
+            join_type=JoinType.LEFT,
+        )
+
+        schema = JoinedSchema(join_structure)
+        dataset_schemas = DatasetSchemas(
+            read_schema=schema,
+            write_schema=None,
+        )
+        super().__init__(
+            dataset_schemas=dataset_schemas,
+            time_group_columns={
+                'events.time': 'events.timestamp',
+            },
+        )
+
+    def default_conditions(self):
+        return [
+            ('events.deleted', '=', 0),
+            ('groups.record_deleted', '=', 0),
+        ]
+
+    def column_expr(self, column_name, body, table_alias: str=""):
+        # Eventually joined dataset should not be represented by the same abstraction
+        # as joinable datasets. That will be easier through the TableStorage abstraciton.
+        # Thus, as of now, receiving a table_alias here is not supported.
+        assert table_alias == "", \
+            "Groups dataset cannot be referenced with table aliases. Alias provided {table_alias}"
+
+        match = self.QUALIFIED_COLUMN_REGEX.match(column_name)
+        if not match:
+            # anything that is not prefixed with a table alias is simply
+            # escaped and returned. It could be a literal.
+            return super().column_expr(column_name, body, table_alias)
+        else:
+            table_alias = match[1]
+            simple_column_name = match[2]
+            if table_alias == self.GROUPS_ALIAS:
+                return self.__grouped_message.column_expr(simple_column_name, body, table_alias)
```

Comment:
With this approach, composite datasets can parse the column name into table alias and column name. 
In theory the events dataset could do this parsing on its own but I opted for not doing it for two reasons:
1) risk: reduce the risk for this PR, so that this parsing logic only happens on a dataset not yet used in prod.
2) going ahead the parsing logic from the column_expr method is going to be separate from the formatting logic and closer to the validation. Finding table aliases will not be a dataset specific functionality but something generic so it would not end into events anyway.

With this approach the events dataset knows table aliases exist, but it does not to make assumptions on the name of the alias.